pkg/eosclient/eosgrpc/eos_http/eoshttp.go

```diff
@@ -0,0 +1,501 @@
+	s := c.opt.BaseURL
+	if len(urlpath) > 0 && urlpath[0] != '/' {
+		s += "/"
+	}
+	s += urlpath
+
+	// I feel safer putting here a check, to prohibit malicious users to
+	// inject a false uid/gid into the url
+	// Who knows, maybe it's redundant? Better more than nothing.
+	p1 := strings.Index(urlpath, "eos.ruid")
+	if p1 > 0 && (urlpath[p1-1] == '&' || urlpath[p1-1] == '?') {
+		return "", errtypes.PermissionDenied("Illegal malicious url " + urlpath)
+	}
+	p1 = strings.Index(urlpath, "eos.guid")
+	if p1 > 0 && (urlpath[p1-1] == '&' || urlpath[p1-1] == '?') {
+		return "", errtypes.PermissionDenied("Illegal malicious url " + urlpath)
+	}
+
+	eosuidgid := ""
+	if len(uid) > 0 {
+		eosuidgid += "eos.ruid=" + uid
+	}
+	if len(gid) > 0 {
+		if len(eosuidgid) > 0 {
+			eosuidgid += "&"
+		}
+		eosuidgid += "eos.rgid=" + gid
+	}
+
+	if strings.Contains(urlpath, "?") {
+		s += "&"
+	} else {
+		s += "?"
+	}
+	s += eosuidgid
+
+	return s, nil
+}
+
+// GETFile does an entire GET to download a full file. Returns a stream to read the content from
+func (c *Client) GETFile(ctx context.Context, remoteuser, uid, gid, urlpath string, stream io.WriteCloser) (io.ReadCloser, error) {
+
+	log := appctx.GetLogger(ctx)
+	log.Info().Str("func", "GETFile").Str("remoteuser", remoteuser).Str("uid,gid", uid+","+gid).Str("path", urlpath).Msg("")
+
+	// Now send the req and see what happens
+	finalurl, err := c.buildFullURL(urlpath, uid, gid)
+	if err != nil {
+		log.Error().Str("func", "GETFile").Str("url", finalurl).Str("err", err.Error()).Msg("can't create request")
+		return nil, err
+	}
+	req, err := http.NewRequestWithContext(ctx, "GET", finalurl, nil)
+	if err != nil {
+		log.Error().Str("func", "GETFile").Str("url", finalurl).Str("err", err.Error()).Msg("can't create request")
+		return nil, err
+	}
+
+	ntries := 0
+	nredirs := 0
+	timebegin := time.Now().Unix()
+
+	for {
+		// Check for a max count of redirections or retries
+
+		// Check for a global timeout in any case
+		tdiff := time.Now().Unix() - timebegin
+		if tdiff > int64(c.opt.OpTimeout) {
+			log.Error().Str("func", "GETFile").Str("url", finalurl).Int64("timeout", tdiff).Int("ntries", ntries).Msg("")
+			return nil, errtypes.InternalError("Timeout with url" + finalurl)
+		}
+
+		// Execute the request. I don't like that there is no explicit timeout or buffer control on the input stream
+		log.Debug().Str("func", "GETFile").Msg("sending req")
+		resp, err := c.cl.Do(req)
+
+		// Let's support redirections... and if we retry we have to retry at the same FST, avoid going back to the MGM
+		if resp != nil && (resp.StatusCode == 307 || resp.StatusCode == 302) {
+
+			// io.Copy(ioutil.Discard, resp.Body)
+			// resp.Body.Close()
+
+			loc, err := resp.Location()
+			if err != nil {
+				log.Error().Str("func", "GETFile").Str("url", finalurl).Str("err", err.Error()).Msg("can't get a new location for a redirection")
+				return nil, err
+			}
+
+			c.cl = &http.Client{
```

Comment:
This will leak a lot of file descriptors. Please remove this and add the `MaxConnsPerHost` and `MaxIdleConnsPerHost` parameters at init

Reply:
Also, the timeout. Uploads/downloads can take a long time

Reply:
> This will leak a lot of file descriptors. Please remove this and add the `MaxConnsPerHost` and `MaxIdleConnsPerHost` parameters at init

It should not, as http clients are pooled/cached in the Transport. I'll add the parms though

Reply:
Then there's no reason to store the client since you're creating a new one every time. So I think you can remove cl from the struct and add transport instead of having it as a global variable